extensions/amp-form/0.1/amp-form.js

```diff
@@ -245,17 +266,113 @@ function onInvalidInputBlur_(event) {
-  // listeners on the same element when the user submit pressing Enter or any
-  // other method to submit the form without the element losing focus.
-  input.removeEventListener('blur', onInvalidInputBlur_);
-  input.removeEventListener('keyup', onInvalidInputKeyUp_);
 
-  input.addEventListener('keyup', onInvalidInputKeyUp_);
-  input.addEventListener('blur', onInvalidInputBlur_);
-  validationBubble.show(input, input.validationMessage);
+/**
+ * Checks user validity for all inputs, fieldsets and the form.
+ * @param {!HTMLFormElement} form
+ */
+function checkUserValidityOnSubmission_(form) {
+  const inputs = form.querySelectorAll('input,select,textarea,fieldset');
+  for (let i = 0; i < inputs.length; i++) {
+    checkUserValidity(inputs[i]);
+  }
+  checkUserValidity(form);
+}
+
+
+/**
+ * Checks user validity which applies .user-valid and .user-invalid AFTER the user
+ * interacts with the input by moving away from the input (blur) or by changing its
+ * value (input).
+ *
+ * See :user-invalid spec for more details:
+ *   https://drafts.csswg.org/selectors-4/#user-pseudos
+ *
+ * The specs are still not fully specified. The current solution tries to follow a common
+ * sense approach for when to apply these classes. As the specs gets clearer, we should
+ * strive to match it as much as possible.
+ *
+ * @param {!HTMLInputElement|!HTMLFormElement|!HTMLFieldSetElement} element
+ * @returns {boolean} true to indicate that validity should propagate to ancestors.
+ */
+function checkUserValidity(element) {
+  const currentUserValid = element.classList.contains('user-valid');
+  const currentUserInvalid = element.classList.contains('user-invalid');
+  if (!currentUserValid && element.checkValidity()) {
+    element.classList.add('user-valid');
+    element.classList.remove('user-invalid');
+    return currentUserInvalid;
+  } else if (!currentUserInvalid && !element.checkValidity()) {
+    element.classList.add('user-invalid');
+    element.classList.remove('user-valid');
+    return true;
+  } else {
+    return false;
+  }
+}
+
+
+/**
+ * Responds to user interaction with an input by checking user validity of the input
+ * and possibly its input-related ancestors (e.g. feildset, form).
+ * @param {!Event} e
+ */
+function onInputInteraction_(e) {
+  const input = e.target;
+  const shouldPropagate = checkUserValidity(input);
+  if (!shouldPropagate) {
+    return;
+  }
+
+  // Propagate user validity to ancestor fieldsets.
+  const ancestors = ancestorElementsByTag(input, 'fieldset');
+  for (let i = 0; i < ancestors.length; i++) {
+    checkUserValidity(ancestors[i]);
+  }
+  // Also update the form user validity.
+  if (input.form) {
+    checkUserValidity(input.form);
+  }
+}
+
+
+/**
+ * Removes event listeners from input.
+ * @param {!Event} e
+ */
+function onInputBlur_(e) {
+  const input = e.target;
+  input.removeEventListener('blur', onInputInteraction_);
```

Comment:
Why do we keep registering/removing these listeners? I doubt it'd hurt to just leave them on the inputs, especially since you're reusing the same function instance.